### ISSUE TITLE
stan_backend is missing when loading model from json

### ISSUE THREAD

**Issue by cherfongfoo** [facebook/prophet#1709]
```
from fbprophet.serialize import model_to_json, model_from_json
with open("serialized_model.json", "r") as fin:
    fb_model = model_from_json(json.load(fin))

df_cv = cross_validation(fb_model, ............)

```

It produces this error,

```
File "C:\Users\XXXXX\.conda\envs\fbprophet7\lib\site-packages\fbprophet\diagnostics.py", line 295, in prophet_copy
    stan_backend=m.stan_backend.get_type()
AttributeError: 'NoneType' object has no attribute 'get_type'

```

Code snippet from diagnostics.py

```
m2 = m.__class__(
        growth=m.growth,
        n_changepoints=m.n_changepoints,
        changepoint_range=m.changepoint_range,
        changepoints=changepoints,
        yearly_seasonality=False,
        weekly_seasonality=False,
        daily_seasonality=False,
        holidays=m.holidays,
        seasonality_mode=m.seasonality_mode,
        seasonality_prior_scale=m.seasonality_prior_scale,
        changepoint_prior_scale=m.changepoint_prior_scale,
        holidays_prior_scale=m.holidays_prior_scale,
        mcmc_samples=m.mcmc_samples,
        interval_width=m.interval_width,
        uncertainty_samples=m.uncertainty_samples,
        stan_backend=m.stan_backend.get_type()
    )
```

It errors out when it tries to copy another model object
I can comment out "stan_backend=m.stan_backend.get_type()"  to make it work, not sure if it has any complication or raise any concern.

Can be easily replicated when you load a model from json, using FBProphet v0.7 on windows.

```
from fbprophet.serialize import model_to_json, model_from_json
with open('serialized_model.json', 'r') as fin:
    fb_model = model_from_json(json.load(fin))
fb_model.stan_backend.get_type()

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-104-5351aed8e4ed> in <module>
----> 1 fb_model.stan_backend.get_type()
 
AttributeError: 'NoneType' object has no attribute 'get_type'
```

**Comment by syedroshanzameer** [comment]
I have the same issue. 
I'm using fbprophet version 0.7.1 with python 3.7.6

`INFO:fbprophet:Making 49 forecasts with cutoffs between 2018-09-11 00:00:00 and 2020-08-31 00:00:00
  0%|          | 0/49 [00:00<?, ?it/s]
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
 in 
      1 ### Cross validation
      2 from fbprophet.diagnostics import cross_validation
----> 3 df_cv = cross_validation(model,horizon= '30 days')

/opt/anaconda3/lib/python3.7/site-packages/fbprophet/diagnostics.py in cross_validation(model, horizon, period, initial, parallel, cutoffs)
    187         predicts = [
    188             single_cutoff_forecast(df, model, cutoff, horizon, predict_columns)
--> 189             for cutoff in tqdm(cutoffs)
    190         ]
    191 

/opt/anaconda3/lib/python3.7/site-packages/fbprophet/diagnostics.py in (.0)
    187         predicts = [
    188             single_cutoff_forecast(df, model, cutoff, horizon, predict_columns)
--> 189             for cutoff in tqdm(cutoffs)
    190         ]
    191 

/opt/anaconda3/lib/python3.7/site-packages/fbprophet/diagnostics.py in single_cutoff_forecast(df, model, cutoff, horizon, predict_columns)
    217 
    218     # Generate new object with copying fitting options
--> 219     m = prophet_copy(model, cutoff)
    220     # Train model
    221     history_c = df[df['ds'] <= cutoff]

/opt/anaconda3/lib/python3.7/site-packages/fbprophet/diagnostics.py in prophet_copy(m, cutoff)
    293         interval_width=m.interval_width,
    294         uncertainty_samples=m.uncertainty_samples,
--> 295         stan_backend=m.stan_backend.get_type()
    296     )
    297     m2.extra_regressors = deepcopy(m.extra_regressors)

AttributeError: 'NoneType' object has no attribute 'get_type'`

**Comment by syedroshanzameer** [comment]
Using joblib module work:

from sklearn.externals import joblib 
  
# Save the model as a pickle in a file 
joblib.dump(model, 'saved_model.pkl') 
  
# Load the model from the file 
knn_from_joblib = joblib.load('filename.pkl')

**Comment by bletham** [comment]
Thanks for raising the issue.

The Stan backend is not included in the json. This was intentional; it's mentioned in the docstring for the function
https://github.com/facebook/prophet/blob/e41ed25646f44f713c110c30c07c678e4a07728e/python/fbprophet/serialize.py#L43
but should probably be described in the documentation too. The focus was to store everything required to make predictions with the model, and so we didn't worry about the stan backend since it isn't.

As you have discovered, cross validation is not possible on a model that has been loaded from a json because of the missing backend. This would be good to fix.

A simple workaround is just to instantiate a new model, and then copy the stan backend from that model over to the model that you've loaded from pickle.

In the longer term, I think this should be a one-line fix: in `prophet_copy`, in this line that is raising the error we should just have it pass along `None` if `m.stan_backend` is `None`:
https://github.com/facebook/prophet/blob/e41ed25646f44f713c110c30c07c678e4a07728e/python/fbprophet/diagnostics.py#L295

**Comment by cherfongfoo** [comment]
> Thanks for raising the issue.
> 
> The Stan backend is not included in the json. This was intentional; it's mentioned in the docstring for the function
> 
> https://github.com/facebook/prophet/blob/e41ed25646f44f713c110c30c07c678e4a07728e/python/fbprophet/serialize.py#L43
> 
> 
> but should probably be described in the documentation too. The focus was to store everything required to make predictions with the model, and so we didn't worry about the stan backend since it isn't.
> As you have discovered, cross validation is not possible on a model that has been loaded from a json because of the missing backend. This would be good to fix.
> 
> A simple workaround is just to instantiate a new model, and then copy the stan backend from that model over to the model that you've loaded from pickle.
> 
> In the longer term, I think this should be a one-line fix: in `prophet_copy`, in this line that is raising the error we should just have it pass along `None` if `m.stan_backend` is `None`:
> 
> https://github.com/facebook/prophet/blob/e41ed25646f44f713c110c30c07c678e4a07728e/python/fbprophet/diagnostics.py#L295

**Comment by cherfongfoo** [comment]
Got it, will close the issue.

**Comment by bletham** [comment]
Actually I think you raised a good issue that we should make the CV work on a model that has been loaded from JSON, and I think it should be possible to do this without having to actually serialize the entire backend object. So I'll leave this open until we get CV working on a loaded model.

**Comment by bletham** [comment]
This is fixed in eb6b8f6. Here is a test case:
```Python
import json
from fbprophet.serialize import model_from_json
from fbprophet.diagnostics import cross_validation


with open('serialized_model_v0.7.1.json', 'rb') as fin:
    model = model_from_json(json.load(fin))

df_cv = cross_validation(model, initial='15 days', horizon='1 day', period='2 days')  # Now works
```

**Comment by bletham** [comment]
The fix has been pushed to pypi in the new v1.0 (with the new package name, prophet)
